capi_problems.rst

```diff
@@ -94,12 +99,100 @@ all types of API users:
 * Manage sub-interpreters
 * Handle and send signals
 
+External users often support multiple Python versions
+or Python implementations. For them, these are the common
+requirements:
+
+* Specification/documentation of the API and ABI
+* Backwards/forwards compatibility and stability for:
+    * API (code works as-is but requires recompilation)
+    * ABI (compiled binaries work)
+    * Across CPython versions
+    * Across Python implementations
+
+Note: in practice, there will be tradeoffs between some of the requirements,
+for example, stability vs. performance.
+
 Groups of users are:
 
+**CPython developers**
+
+CPython developers change the CPython internals to implement new
+functionality or improve/optimize the current functionality. While ideally,
+they would be free to alter any internal details, in practice, anything that
+is intentionally (or even unintentionally) exposed to external C API users
+should remain compatible (depending on the concrete policy for the
+specific API). If it is part of the Stable ABI, it must remain binary
+compatible.
+
+One recent notable example is that a significant amount of work on the
+Gilectomy project and the `PEP 703 <https://peps.python.org/pep-0703/>`__
+was spent on dealing with reference counting, most specifically on
+keeping the reference counting contract, which is only exposed in C API,
+while making it scale with parallel execution.
+
+For another example, see the Discuss
+`thread <https://discuss.python.org/t/lets-get-rid-of-the-stable-abi-but-keep-the-limited-api/18458>`__:
+Let’s get rid of the stable ABI, but keep the limited API.
+
+Because the CPython code is by definition intended to run only on given
+version of CPython and is always recompiled, the following requirements
+*do not apply* to the C API usage in CPython core codebase:
+
+* Compatibility with alternative Python implementations
```

Comment:
Sure, that's the public contract of Python, but from the point of view of CPython developer I don't think that it matters whether the API that is used to implement CPython itself supports alternative Pythons or not (and should not be visible in Python's public contract). For alternative Pythons it matters whether the C API that extensions use is implementable with their internal structure. Those two APIs (for CPython core, for 3rd party extensions) are one thing currently (or overlap a lot), but they do not have to be.

Maybe it should clarify that in the section "CPython developers" we focus on the development of CPython core and the C API's requirements for that purpose only. If those requirements are in contradiction with requirements for other use-cases, it's fine. I think that's the point of collecting the requirements.

Reply:
Is even correct to say that CPython Core (excluding stdlib extensions) is using the cpython API? It can access whatever it wants.

Reply:
Perhaps we should just focus on the requirements that different stakeholders have, and omit statements about what requirements they don't care about.  So rather than explaining how python core doesn't need to support multiple versions, we mention that many extension writers do.

Reply:
This is a good point. The document actually does not specify what exactly is "C API" in its context. Is it limited API, things exposed in `Python.h`, things exposed in any header in Python's `include` directory? In any case, those header files and the functions declared in them are shared between external code (extensions, embedders) and the interpreter itself. Also the introduction of this document mentions:
> It [C API] evolved from what was initially the internal API between the C code of the interpreter and the Python language and libraries

I think that it is a real issue that the current C API (even if we think of it as the most restricted version: the limited API) hinders the development of the internals of CPython itself. The linked discussion demonstrates that clearly I believe. There is also discussion about stable ABI and PEP 703, because it needs to hide implementation details of reference counting behind actual function call as opposed to C macros, i.e., break the stable ABI. I think all that is evidence that there should be a stakeholder section for CPython and it should list those issues.

My specific point here is that whatever CPython uses to implement itself, does not actually need to work for other Pythons and does not need any stability guarantees. Whether that is the same API, some superset, or something completely different belongs to the design/implementation of whatever comes up from assessing the requirements described here.

Admittedly, I believe that C API for 3rd party extensions that hides the implementation details enough to be really future proof not only for other Pythons but future development of CPython, must be inherently different from what CPython uses internally. The fact that the current C API can somehow serve both as public C API for external code and internal API actually shows that it is not an abstract enough. However, this is a conclusion/solution, which is what I am trying to avoid in this document, and I am rather trying to describe the problems that lead me to this conclusion :-)

I will think about rephrasing this section bit more. Any suggestions are welcome :-)